File: JointMatrixFactorization-Scala/src/main/scala/de/tu_berlin/dima/recommendationsystem/InitItemFeatureVectorReducer.scala
package de.tu_berlin.dima.recommendationsystem

import org.apache.mahout.math.RandomAccessSparseVector;
import org.apache.mahout.math.SequentialAccessSparseVector;
import org.apache.mahout.math.Vector;
import org.apache.mahout.math.als.AlternatingLeastSquaresSolver;
import org.apache.mahout.math.map.OpenIntObjectHashMap;

import eu.stratosphere.api.scala.functions._
import java.util.Random;

class InitItemFeatureVectorReducer extends GroupReduceFunction[(Int, Int, Float), (Int, PactVector)] {
	override def apply (in: Iterator[(Int, Int, Float)]) : (Int, PactVector) = {
	  val numfeatures = Util.numFeatures
	  val random = new Random()
	  val featureVector = new PactVector()
	  val features : Vector = new SequentialAccessSparseVector(Integer.MAX_VALUE, Util.numFeatures)
	  val itemID = in.next._2;
	  for (i <- 1 to numfeatures) {
	    val t = random.nextFloat
	    features.setQuick(i - 1, t)
	  }
	  featureVector.set(features)
	  (itemID, featureVector)
	}
}

File: jointmatrixfactorization/src/main/java/de/tu_berlin/dima/bigdata/jointmatrixfactorization/mapper/InitItemFeatureVectorReducer.java
package de.tu_berlin.dima.bigdata.jointmatrixfactorization.mapper;

import java.util.Iterator;
import java.util.Random;

import org.apache.mahout.math.SequentialAccessSparseVector;
import org.apache.mahout.math.Vector;

import de.tu_berlin.dima.bigdata.jointmatrixfactorization.type.PactVector;
import de.tu_berlin.dima.bigdata.jointmatrixfactorization.util.Util;
import eu.stratosphere.pact.common.stubs.Collector;
import eu.stratosphere.pact.common.stubs.ReduceStub;
import eu.stratosphere.pact.common.type.PactRecord;
import eu.stratosphere.pact.common.type.base.PactInteger;

/**
 * Input: <userID, itemID, Rating> from TuppleMapper
 * Reduce Key: itemID
 * Output: <itemID, initFeatureVector>
 * 
 * @author titicaca
 *
 */

public class InitItemFeatureVectorReducer extends ReduceStub{

	private final Vector features = new SequentialAccessSparseVector(Integer.MAX_VALUE, Util.numFeatures);
	private final Random random = new Random();
	private final PactRecord outputRecord = new PactRecord();
	private final int numFeatures = Util.numFeatures;
	private final PactVector featureVector = new PactVector();

	
	@Override
	public void reduce(Iterator<PactRecord> records, Collector<PactRecord> collector)
			throws Exception {

		int itemID = -1;
		if(records.hasNext()){
			itemID = records.next().getField(1, PactInteger.class).getValue();	
		}
		if(itemID > 0){
			for(int i = 0; i < numFeatures; i ++){
				features.set(i, random.nextFloat());
			}
			featureVector.set(features);
			
			outputRecord.setField(0, new PactInteger(itemID));
			outputRecord.setField(1, featureVector);
			
			collector.collect(outputRecord);
		}
		
	}
	
}